macrosynergy/pnl/proxy_pnl_calc.py

```diff
@@ -7,31 +7,174 @@
 import matplotlib.pyplot as plt
 import seaborn as sns
 
-from typing import List, Union, Tuple, Optional
+from typing import List, Union, Tuple, Optional, Dict, Callable
+from numbers import Number
 
-from macrosynergy.management.simulate import make_qdf
-from macrosynergy.management.utils import reduce_df
+from macrosynergy.management.simulate import make_test_df
+from macrosynergy.download.transaction_costs import download_transaction_costs
+from macrosynergy.management.utils import (
+    reduce_df,
+    get_cid,
+    get_xcat,
+    ticker_df_to_qdf,
+    qdf_to_ticker_df,
+)
+from macrosynergy.management.types import QuantamentalDataFrame
+from macrosynergy.pnl.transaction_costs import TransactionCosts
+
+
+def get_diff_index(df_wide: pd.DataFrame) -> pd.Index:
+    df_diff = df_wide.diff(axis=0)
+    change_index = df_diff.index[(df_diff.abs() > 0).any(axis=1)]
+    return change_index
+
+
+def _replace_strs(
+    list_of_strs: List[str], old_str: str, new_str: str = ""
+) -> List[str]:
+    return [_.replace(old_str, new_str) for _ in list_of_strs]
+
+
+def _prep_dfs_for_pnl_calcs(
+    df: QuantamentalDataFrame,
+    spos: str,
+    rstring: str,
+) -> Tuple[pd.DataFrame, pd.DataFrame, pd.DataFrame, List[pd.Timestamp]]:
+    df_wide = qdf_to_ticker_df(df=df.copy())
+
+    # Filter
+    u_tickers: List[str] = list(set(df["cid"] + "_" + df["xcat"]))
+    filt_xrs: List[str] = [tx for tx in u_tickers if tx.endswith(rstring)]
+    # do xrs contain `spos` at this point?
+    filt_pos = [tx for tx in u_tickers if tx.endswith(spos)]
+
+    assert set(_replace_strs(filt_xrs, rstring)) == set(
+        _replace_strs(filt_pos, f"_{spos}")
+    )
+
+    # Pivot the dataframes
+    pivot_returns: pd.DataFrame = df_wide.loc[:, filt_xrs]
+    pivot_pos: pd.DataFrame = df_wide.loc[:, filt_pos]
+
+    # warn about NAs
+    dfx: pd.DataFrame
+    for dfx, dfname in [(pivot_returns, "returns"), (pivot_pos, "positions")]:
+        # for each column warns for dates of nas
+        for col in dfx.columns:
+            nas_idx = dfx[col].loc[dfx[col].isna()]
+            if not nas_idx.empty:
+                print(
+                    f"Warning: Series {col} has NAs at the following dates: {nas_idx.index}"
+                )
+
+    # Get the diff index for positions
+    pos_diff_index: pd.DatetimeIndex = get_diff_index(df_wide=pivot_pos)
+    start: str = pivot_pos.first_valid_index()
+    end: str = pivot_pos.last_valid_index()
+
+    # List of rebal_dates - with end date as the last date of the PNL calc
+
+    rebal_dates = sorted(set(pos_diff_index) | {pd.Timestamp(end)})
```

Comment:
Why are we rebalancing on the end date? If it isn't on the first, we shouldn't rebalance.

Reply:
The dnd date will end up as d2, so we're rebalancing up to but not including the end day.

Reply:
A better way to put it would be:
On the last rebalance day, as took a position, which continues to have a PNL until no more position data is available

Reply:
Got it, that makes more sense. Thank you. Is it maybe worth to add the end-date only in the function when we rebalance (`rebal_dates + [pd.Timestamp(end)]`) to make it clear that `rebal_dates` are dates of rebalancing.

Reply:
does make sense, changing now